collections/collections.go

```diff
@@ -0,0 +1,98 @@
+package collections
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// ErrNotFound defines an error returned when
+// a key was not found.
+type ErrNotFound struct {
+	// HumanizedKey represents the human-readable key.
+	HumanizedKey string
+	// RawKey represents the raw key in byte format.
+	RawKey []byte
+	// ValueType represents the type which was not found.
+	ValueType string
+}
+
+// Error implements the error interface.
+func (e ErrNotFound) Error() string {
+	return fmt.Sprintf("collections: key not found for %s '%s'", e.ValueType, e.HumanizedKey)
+}
+
+// Is implements error.Is interface
+func (e ErrNotFound) Is(err error) bool {
+	self, ok := err.(ErrNotFound)
+	if !ok {
+		return false
+	}
+	return bytes.Equal(self.RawKey, e.RawKey) &&
+		self.HumanizedKey == e.HumanizedKey &&
+		self.ValueType == e.ValueType
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		prefix = c // maybe copy?
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyEncoder defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyEncoder[T any] interface {
+	// Encode encodes the type T into bytes.
+	Encode(key T) ([]byte, error)
+	// Decode decodes the given bytes back into T.
+	// And it also must return the bytes of the buffer which were read.
+	Decode(b []byte) (int, T, error)
```

Comment:
```suggestion
	// Encode encodes the type T into bytes as a standalone single key or the terminal part of a multi-part key.
	Encode(key T) ([]byte, error)
	// Decode decodes the given bytes back into T.
	// And it also must return the bytes of the buffer which were read.
	Decode(b []byte) (int, T, error)
	// EncodeNonTerminal encodes the key as a non-terminal part of a multi-part key.
	EncodeNonTerminal(key T) ([]byte, error)
	DecodeNonTerminal(b []byte) (int, T, error)
```

Can we separate terminal and non-terminal encodings? This is something that we do in the SDK and is also done in the ORM for multi-part keys, so would help for compatibility as well as correctness.

A specific case example would be a pair of `[]byte` keys. The first `[]byte` key needs to be length-prefixed, but the second `[]byte` part does not need the length prefix because we can assume all the remaining bytes are part of it. For string keys, we need a null-terminator for non-terminal segments but we don't need it for the terminal segment. Having these separated both saves space and also ensure keys are properly implemented to be used in multi-part keys.

Reply:
Also, a couple other things I would bring over from the ORM are the sizing methods: https://github.com/cosmos/cosmos-sdk/blob/ca0b1deeea101fb0fcdc69cab98a120b8baa3241/orm/encoding/ormfield/codec.go#L30-L37 and using `io.Reader` and `io.Writer` for Encode/Decode. This allows a single buffer to be allocated in multi-part keys.

Reply:
> Can we separate terminal and non-terminal encodings? 

Yes this is something that I foresaw, but I didn't put it in KeyEncoder, because if it exposed there it means it's `collections` which decides when to pick the NullTermination path, which would not be correct to retain backwards compat.

My thinking was to have a separate class of KeyEncoders eg: 
NullTerminated for types: ~string
LengthPrefixed for types: ~[]byte (AccAddress, ValAddress,...)

> Also, a couple other things I would bring over from the ORM are using io.Reader and io.Writer in Encode/Decode and the sizing methods

Sizing methods sound good to me in order to optimise allocation, what about adding a `Size(K) int` method?

Constant size encoders would implement it as (uint64 example) `Size(key _) int { return 8 }`, others would return like `len(key)` etc.

WRT `io.Reader\Writer` wouldn't it be more performant to use a byte slice? I could see advantage of the Reader\Writer pattern only if keys need to be read\written in batch from external consumers, but then I'd rather have a StreamKeyEncoder wrapper for this specific use-case.

Reply:
> Yes this is something that I foresaw, but I didn't put it in KeyEncoder, because if it exposed there it means it's collections which decides when to pick the NullTermination path, which would not be correct to retain backwards compat.

It's `Pair`, `Triple`, etc. which should decide what is terminal/non-terminal IMHO. If you need backwards compatibility with existing code you have, you just need a set of encoders that do the same in both thing for terminal/non-terminal. But I think going forward, it would be good to separate and then let `Pair`, etc. call the correct method. We actually need that behavior for ORM and some SDK keys compatibility.

> Sizing methods sound good to me in order to optimise allocation, what about adding a Size(K) int method?
> 
> Constant size encoders would implement it as (uint64 example) Size(_ uint64) int { return 8 }, others would return like len(key) etc.

Yep makes sense. Varints in the ORM implement it as the max varint size - it's okay to over-allocate a little bit. The reason for the `FixedSize` method is that the multi-part encoder will cache the size for parts that are always fixed: https://github.com/cosmos/cosmos-sdk/blob/ca0b1deeea101fb0fcdc69cab98a120b8baa3241/orm/encoding/ormkv/key_codec.go#L69-L70. Maybe that's over-optimizing but it doesn't hurt.